### PR TITLE
Ordered polls

### PR DESCRIPTION
**Pull request by javiexin**
These are my changes for ordered polls.
NOTE: It requires a new event in viewtopic, included the (complete) viewtopic file.
NOTE: The combined behaviour of both your ext and mine is bad, test it please.
NOTE: I have made no JS changes, they might be required, but it goes over my level at this point.

Let's discuss when you review the code and test it.  Config is OK, posting is OK, viewtopic needs to be reviewed thoroughly.

Thanks,
-javiexin

### CODE REVIEW

**Comment by Wolfsblvt** [comment]
Take a look a the failing travis tests: https://travis-ci.org/Wolfsblvt/advancedpolls/builds/57416010

First one is cause of the `viewtopic.php` file you have included. That's a bad idea. It is not part of the extension, so validator will mess up and it has no point in the repository. If you just want to give me the code for the event, another thing like a pastebin and a link would be better. Or best: A pull request for that event to phpBB core. A pull request link here would mean I just can pull the pull request locally and test it with it there.

Second one is this:

    FILE: ...s/build/phpBB3/phpBB/ext/wolfsblvt/advancedpolls/core/advancedpolls.php
    --------------------------------------------------------------------------------
    FOUND 1 ERROR(S) AFFECTING 1 LINE(S)
    --------------------------------------------------------------------------------
     368 | ERROR | Whitespace found at end of line
         |       | (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)

You can just fix this by removing tabs and spaces from the end of that line. (Empty lines should not contain any whitespace at all)

File: core/advancedpolls.php [hunk 1]
```diff
@@ -346,6 +352,23 @@ public function do_poll_modification($topic_data)
 	}
 
 	/**
+	 * Add the possible options to the template
+	 *
+	 * @param array	$post_data		The array of post data
+	 * @return void
+	 */
```
Comment (Wolfsblvt) [inline core/advancedpolls.php]: Please update this docblock with a short explaining sentence what this function does/what it is for and the correct parameter name, data type and description.

That helps me to understand the code months and years later :P And my IDE also shows the text on IntelliSense.

Ah, and while you are at it. Even an internal function like the `order_by_votes()` below should have a docblock. parameters of course, and maybe a sentence like "Function used for sorting blah bblahb blah"

File: event/listener.php [hunk 2]
```diff
@@ -111,6 +112,22 @@ public function do_poll_modification($event)
...
 
 	/**
+	 * Modifys the template vars to match the advanced poll settings
+	 *
+	 * @param object $event The event object
+	 * @return void
+	 */
+	public function do_poll_block_modification($event)
+	{
+		if ($event['s_display_results'] && (int) $event['topic_data']['wolfsblvt_poll_show_ordered'])
```
Comment (Wolfsblvt) [inline event/listener.php]: Sure that `$event['s_display_results']` is the best idea? I mean if you haven't voted, this function will not run. But if you vote then via AJAX, the votes are not ordered.
So why not use the same idea like mine, check for `isset($topic_data['poll_title'])` there?
Oh, and the int cast is not needed I would say. If you want typecasting, `(bool)` would be the correct here.

File: core/advancedpolls.php [hunk 3]
```diff
@@ -346,6 +352,23 @@ public function do_poll_modification($topic_data)
...
+	public function do_poll_block_modification($poll_options_template_data)
+	{
+		usort($poll_options_template_data, array($this, "order_by_votes"));
+		return $poll_options_template_data;
+	}
+
+	protected function order_by_votes($a, $b)
+	{
+		return (((int) $a['POLL_OPTION_RESULT'] < (int) $b['POLL_OPTION_RESULT']) ? 1 : (((int) $a['POLL_OPTION_RESULT'] > (int) $b['POLL_OPTION_RESULT']) ? -1 : (((int) $a['POLL_OPTION_ID'] > (int) $b['POLL_OPTION_ID']) ? 1 : -1)));
```
Comment (Wolfsblvt) [inline core/advancedpolls.php]: Easier statement would be:
```php
return ((int) $b['POLL_OPTION_RESULT'] - (int) $a['POLL_OPTION_RESULT']))
```
For usort the it doesn't matter if you return 1 or 5, that just means `$b` is higher.

**Comment by javiexin** [comment]
Ok, I will remove the viewtopic file from the repo and pull request (if I
know how :)).  As for the phpbb pull request, here it is:
https://github.com/phpbb/phpbb/pull/3521

I will work on the other comments/errors and resubmit.

Thanks a lot,
-javiexin

PS: Bear with me, I haven't used github ever before...

2015-04-07 3:30 GMT+02:00 Wolfsblvt <notifications@github.com>:

**Comment by Wolfsblvt** [comment]
You can remove like you edit them now in your local repo where you work on. When you do another commit, the pull request here will be updated automatically.

Thank you for your help (:

File: core/advancedpolls.php (hunk 1 above)
Comment (javiexin) [inline core/advancedpolls.php]: Ok, I will.  Take a look at the other docblocks in the file (yours): one is
misplaced (duplicate of another), and another is a copy&paste of another
ext.

2015-04-07 3:46 GMT+02:00 Wolfsblvt <notifications@github.com>:

File: core/advancedpolls.php (hunk 3 above)
Comment (javiexin) [inline core/advancedpolls.php]: Not so much simpler your way, should be:

return (((int) $b['POLL_OPTION_RESULT'] != (int) $a['POLL_OPTION_RESULT'])
? ((int) $b['POLL_OPTION_RESULT'] - (int) $a['POLL_OPTION_RESULT']) :
((int) $a['POLL_OPTION_ID'] - (int) $b['POLL_OPTION_ID']) );

In case of same number of votes, the order should be by option id.


2015-04-07 4:08 GMT+02:00 Wolfsblvt <notifications@github.com>:

File: event/listener.php (hunk 2 above)
Comment (javiexin) [inline event/listener.php]: This is intentional: poll should not be ordered unless RESULTS are
displayed, if you have not voted, order should be by option id (so that you
are not influenced).  So no, it should not run if you have not voted, or if
the results are hidden.  I will verify the coding anyway, as I have to look
carefully at the conditions.

Ah, AJAX... that is the one bit that I don't know how to make it work, the
JS part that I mentioned.  Any suggestion on how to address this?

Regarding the type casting, ok, no big deal :).

2015-04-07 3:59 GMT+02:00 Wolfsblvt <notifications@github.com>:

File: core/advancedpolls.php (hunk 3 above)
Comment (Wolfsblvt) [inline core/advancedpolls.php]: Ah I see. usort is not stable, that's the problem.

In this case, you can use
```php
return ((int) $b['POLL_OPTION_RESULT'] - (int) $a['POLL_OPTION_RESULT'])) ?: ((int) $a['POLL_OPTION_ID'] - (int) $b['POLL_OPTION_ID']));
```
This is the shorthand [ternary operator](http://php.net/manual/en/language.operators.comparison.php#language.operators.comparison.ternary). See [here on stackoverflow](http://stackoverflow.com/questions/1993409/operator-the-elvis-operator-in-php) for an explanation.

But I would guess in this case I would go another way. Make a real if/else statement. That will be better readable.

File: event/listener.php (hunk 2 above)
Comment (Wolfsblvt) [inline event/listener.php]: Okay, in this case we have to sort manually after giving the vote via ajax. Extending the vote function there.
You can create an issue for that, than I will remember it that I have to do it for your stuff.

File: core/advancedpolls.php (hunk 3 above)
Comment (javiexin) [inline core/advancedpolls.php]: Great, thanks, I didn't know about the ternary operator used that way.  And for the general case, I would do as you suggest, go for the readability.  But for a sorting function that is called numerous times, and is not complex to understan, I prefer the ternary op...

Comment (Wolfsblvt) [inline core/advancedpolls.php]: You can do that of course, it is understandable.
It wouldn't make any difference though, internally ternary operator will be executed the same way like an if/else statement :P

File: event/listener.php (hunk 2 above)
Comment (javiexin) [inline event/listener.php]: I have looked into this further, and I have pushed this check into the advancedpolls class, while I keep here the same check that you have for the other cases, as you suggested.
The issue is already created.

**Comment by Wolfsblvt** [comment]
Can you keep me updated what you have done etc?
I see many commits here :P
Oh, and merging things from another pull request in this pull request is maybe not the best idea. Now you have everything here and we may just close the other one.

It is better to keep things seperated, if possible. This PR gets a bit.. overloaded :P

**Comment by javiexin** [comment]
Sure.
1) On the merge, I sent you the changes that fixed the mentioned issues separately because they applied directly to your current version, and may be incorporated immediately.  Then, I merged the changes into this branch, and, as far as I recall, did not push it again, thinking this would not be sent to you...
2) On progress and commits in this branch, first commits refer to the main functionality of the enhancement, some small bug fixing... then I applied the issue fixes before I talked to you, when you told me to separate them into a different pull request, so I reverted that commit.  And finally, I moved all of your code from your current event to the new one, so that everything is more homogeneous, and ready for the next round.  Then re-merged the fixes (and didn't push that commit, but seems to have been uploaded anyway).

I think the enhancement is ready for thorough testing, other that that completed, including the move of all of your code to the new event.

Now I am working on the next set of functionality, and I will make sure it comes in a separate branch and pull request, to be incorporated later.  Hopefully, this time more successfully, as I am getting more experienced with GIThub  :(

**Comment by Wolfsblvt** [comment]
Okay, I understand. Travis tells me code is correct. I have no unit tests, so I think I will test it a bit myself but I am quite sure it will work, I trust you there :P

So this pull request is finished and ready to merge?
if you have finished the work, I will see if I have comments on your code.

**Comment by javiexin** [comment]
Yes, I would consider this finished, unless some bug comes up, or there are suggestions on improvements...  Only missing is the JS update after voting (sorting and removing label), but I am not positive this is needed at all (beyond what you already have).

**Comment by Wolfsblvt** [comment]
If you can make it best, why not do it? :P JS shouldn't be a big problem. I will look on the pull request when I have time

**Comment by javiexin** [comment]
Just so you know, the event this relies on has been moved to 3.1.5.  So it won't be available in 3.1.4. You may see the discussion here: https://github.com/phpbb/phpbb/pull/3521
